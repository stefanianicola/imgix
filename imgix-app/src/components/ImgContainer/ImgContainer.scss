.img-content{
    padding: 0;
    cursor: pointer;
    opacity: 0.7;
    width: 160px;
    height: 150px;
    &:hover{
        opacity: 1;
    }
}