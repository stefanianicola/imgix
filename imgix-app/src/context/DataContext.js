import React, { createContext, useState, useEffect } from 'react';
import axios from 'axios';
import Params from '../Data/Params';
export const DataContext = createContext();


const DataContextProvider = (props) => {
    const [data, setData] = useState([]);
    const [flag, setFlag] = useState(false);
    const [btn, setBtn] = useState([]);
    const [nameButtons, setNameButtons] = useState([]);
    const [loading, setLoading] = useState(false);
    const [params, setParams] = useState([]);
    const [url, setUrl] = useState("");
    const [hideMenu, setHideMenu] = useState(false);

    const urlImg = "https://assets.imgix.net/unsplash/";

    //consume photos API
    useEffect(() => {
        setLoading(true);
        try {
            axios("https://storage.googleapis.com/nanlabs-engineering-technical-interviews/imgix-samples-list.json")
                .then((res) => {
                    setData(res.data);
                })
        } catch (error) {
            console.log(error);
        }
        setLoading(false);
    }, [])

    //Pass data from the params to the state of btn so that the names appear in the buttons
    useEffect(() => {
        setNameButtons(Params);
    }, [])


    //Pass the parameters of the button that you click
    const addControl = (v) => {
        let parametros = [...Params]
        parametros.forEach((param) => {
            if (v === param.alias) {
                setBtn(param)
            }
        })
        setFlag(true);
    }

    // filter the data of the params to be able to take a single value per parameter
    // and build the url with all the effects we want in the photos
    const addParamUrl = (alias, value) => {
        let aux = [...params];
        let isParam = false;

        aux.forEach((a) => {
            if (a.alias === alias) {
                isParam = true;
                a.value = value;
                setParams(aux);
            }
        })
        !isParam &&
            setParams([...aux, { alias, value }]);
        convertUrl();

    }

    //Convert the array of objects into a string to use as the url of the image
    const convertUrl = () => {
        let stringP = "";
        for (let i = 0; i < params.length; i++) {
            stringP = stringP + `${params[i].alias}=${params[i].value}&`

        }
        setUrl(stringP);

    }

    //reset url with change img
    const resetUrl = () => {
        setUrl("");
        setParams([]);
        setFlag(false);
    }

    //delete the last parameter added
    const removeItem = () => {
        params.pop();
        convertUrl();
        setFlag(false);
    }

    return (
        <DataContext.Provider
            value={{
                data,
                setData,
                addControl,
                nameButtons,
                flag,
                btn,
                loading,
                params,
                setParams,
                addParamUrl,
                url,
                removeItem,
                hideMenu,
                setHideMenu,
                resetUrl,
                urlImg
            }}>
            {props.children}
        </DataContext.Provider>
    )
}

export default DataContextProvider;
