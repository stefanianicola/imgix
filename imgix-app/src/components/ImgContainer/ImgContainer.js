import React, { useContext } from 'react';
import './ImgContainer.scss';
import { Row, Col, Image } from 'react-bootstrap';
import 'bootstrap/dist/css/bootstrap.min.css';
import { DataContext } from '../../context/DataContext';
import { Link } from 'react-router-dom';



const ImgContainer = () => {
    const { data } = useContext(DataContext);

    return (
        <div>
            <Row className="justify-content-center m-0 p-0">
           
                {
                    data.map((d, i) => {
                        return (
                            <Col xs={12} md={2} className="img-content" key={i}>
                                <Link to={`/${d.name}`}>
                                    <Image src={d.url}
                                        alt=""
                                        width={160}
                                        height={150}
                                    />
                                </Link>
                            </Col>
                        )
                    })
                }
            </Row>
        </div>
    )
}
export default ImgContainer;