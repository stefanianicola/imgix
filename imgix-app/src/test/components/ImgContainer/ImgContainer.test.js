import React from 'react';
import { BrowserRouter as Router } from 'react-router-dom';
import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import ImgContainer from '../../../components/ImgContainer/ImgContainer'
import { DataContext } from '../../../context/DataContext';

describe('ImgContainer component', () => {
    const data = [{
        "url": "https://assets.imgix.net/unsplash/alarmclock.jpg",
        "name": "alarmclock.jpg"
    }]
    const resetUrl = jest.fn();
    const wrapper = mount(
        <Router>
            <DataContext.Provider value={{ data, resetUrl }} >
                <ImgContainer />
            </DataContext.Provider>
        </Router>
    );


    it('should be create the component', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should eject resetUrl function', () => {
        const link = wrapper.find('a');
        link.simulate('click');

        expect(resetUrl).toHaveBeenCalled();
    });

})