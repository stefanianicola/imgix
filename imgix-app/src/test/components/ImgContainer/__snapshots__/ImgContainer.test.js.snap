// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImgContainer component should be create the component 1`] = `
<div>
  <div
    className="justify-content-center m-0 p-0 row"
  >
    <div
      className="img-content col-md-2 col-12"
    >
      <a
        href="/alarmclock.jpg"
        onClick={[Function]}
      >
        <img
          height={220}
          src="https://assets.imgix.net/unsplash/alarmclock.jpg?auto=format&ixlib=react-9.1.1&h=220&w=230"
          srcSet="https://assets.imgix.net/unsplash/alarmclock.jpg?auto=format&ixlib=react-9.1.1&w=230&h=220&dpr=1&q=75 1x,
https://assets.imgix.net/unsplash/alarmclock.jpg?auto=format&ixlib=react-9.1.1&w=230&h=220&dpr=2&q=50 2x,
https://assets.imgix.net/unsplash/alarmclock.jpg?auto=format&ixlib=react-9.1.1&w=230&h=220&dpr=3&q=35 3x,
https://assets.imgix.net/unsplash/alarmclock.jpg?auto=format&ixlib=react-9.1.1&w=230&h=220&dpr=4&q=23 4x,
https://assets.imgix.net/unsplash/alarmclock.jpg?auto=format&ixlib=react-9.1.1&w=230&h=220&dpr=5&q=20 5x"
          width={230}
        />
      </a>
    </div>
  </div>
</div>
`;
